### PR TITLE
Add goimports functionality to DocumentFormatting

### LINKED ISSUE #272: Add goimports functionality to the DocumentFormatting
Not sure if this is something you'd consider, but it would be nice if it was possible to auto add missing imports.

I use this diff locally, which seems to do the trick, but I haven't tested this thoroughly yet:

```diff
diff --git a/langserver/format.go b/langserver/format.go
index 7678fc9..6ab56bd 100644
--- a/langserver/format.go
+++ b/langserver/format.go
@@ -4,13 +4,13 @@ import (
        "bytes"
        "context"
        "fmt"
-       "go/ast"
        "go/parser"
        "go/printer"
        "go/token"
        "path"
 
        "golang.org/x/tools/go/buildutil"
+       "golang.org/x/tools/imports"
 
        "github.com/sourcegraph/go-langserver/langserver/util"
        "github.com/sourcegraph/go-langserver/pkg/lsp"
@@ -33,8 +33,6 @@ func (h *LangHandler) handleTextDocumentFormatting(ctx context.Context, conn jso
                return nil, err
        }
 
-       ast.SortImports(fset, file)
-
        var buf bytes.Buffer
        cfg := printer.Config{Mode: printer.UseSpaces | printer.TabIndent, Tabwidth: 8}
        err = cfg.Fprint(&buf, fset, file)
@@ -43,6 +41,12 @@ func (h *LangHandler) handleTextDocumentFormatting(ctx context.Context, conn jso
        }
 
        b := buf.Bytes()
+
+       b, err = imports.Process("", b, nil)
+       if err != nil {
+               return nil, err
+       }
+
        orig, err := h.readFile(ctx, params.TextDocument.URI)
        if err != nil {
                return nil, err
```

### PR DESCRIPTION
**Pull request by otto-md**
Original code by @JeanMertz, some tweaks by me.

Fixes #272

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,6 +42,20 @@ interface GoInitializationOptions {
...
+   * goimportsEnabled formats documents using goimports instead of normal go fmt.
+   *
+   * Defaults to true if not specified.
+   */
+  goimportsEnabled?: boolean;
+
+  /**
+   * goimportsLocalPrefix sets the local prefix (comma-separated string) that goimports will use.
+   *
+   * Defaults to emptry string if not specified.
```
Comment (lloiser) [inline README.md]: typo: empty

File: langserver/config.go [hunk 2]
```diff
@@ -29,6 +29,16 @@ type Config struct {
...
 	GocodeCompletionEnabled bool
 
+	// GoimportsEnabled formats documents using goimports instead of normal go fmt
+	//
+	// Defaults to true if not specified.
+	GoimportsEnabled bool
+
+	// GoimportsLocalPrefix sets the local prefix (comma-separated string) that goimports will use
+	//
+	// Defaults to emptry string if not specified.
```
Comment (lloiser) [inline langserver/config.go]: typo: empty

File: langserver/format.go [hunk 3]
```diff
@@ -45,6 +48,15 @@ func (h *LangHandler) handleTextDocumentFormatting(ctx context.Context, conn jso
 	}
 
 	formatted := buf.Bytes()
+
+	if h.config.GoimportsEnabled {
+		imports.LocalPrefix = h.config.GoimportsLocalPrefix
+		formatted, err = imports.Process("", formatted, nil)
```
Comment (lloiser) [inline langserver/format.go]: So the file is actually formatted twice using `cfg.Print` (aka `gofmt`) and `imports.Process` (aka `goimports`). Seems unnecessary in my opinion...

In my opinion the code could look something like this:
```go
if h.config.GoimportsEnabled {
    // set LocalPrefix
    // imports.Process(...)
} else {
    // ast.SortImports(...)
    // cfg.FPrint(...)
}
```

Comment (otto-md) [reply langserver/format.go]: Yeah, I thought about this too. I have changed it now. It wasn't completely obvious to me how to do it though, but I ended up changing the `imports.Process()` call slightly, it seems to work.

**Comment by keegancsmith** [comment]
Windows CI is not happy with this change. Note master branch is passing.

```
        --- FAIL: TestServer/go_basic/formatting-a.go (0.00s)
        	langserver_test.go:1707: jsonrpc2: code 0 message: open /src/test/pkg/a.go: The system cannot find the path specified.
```

File: langserver/format.go [hunk 4]
```diff
@@ -28,23 +29,34 @@ func (h *LangHandler) handleTextDocumentFormatting(ctx context.Context, conn jso
...
-	if err != nil {
-		return nil, err
-	}
 
-	ast.SortImports(fset, file)
+	var formatted []byte
+	if h.config.GoimportsEnabled {
+		imports.LocalPrefix = h.config.GoimportsLocalPrefix
+		var err error
+		formatted, err = imports.Process(filename, nil, nil)
```
Comment (lloiser) [inline langserver/format.go]: The go-langserver maintains its own file cache, like all other LSP servers do.
So by passing `nil` as second parameter to `Process` forces the imports code to actually read the file from disk without trying to read from the internal file cache!
Just move the `unformatted` variable before this `if` and pass the `unformatted` as second argument to `Process`

**Comment by otto-md** [comment]
Unfortunately I dont know what could cause the Windows CI issue and I dont have access to any windows computers.

File: langserver/format.go [hunk 5]
```diff
@@ -28,27 +29,38 @@ func (h *LangHandler) handleTextDocumentFormatting(ctx context.Context, conn jso
...
 	if err != nil {
 		return nil, err
 	}
 
-	ast.SortImports(fset, file)
+	var formatted []byte
+	if h.config.GoimportsEnabled {
+		imports.LocalPrefix = h.config.GoimportsLocalPrefix
+		var err error
+		formatted, err = imports.Process("", unformatted, nil)
```
Comment (lloiser) [inline langserver/format.go]: pass `filename` instead of `""` for the first argument. This should fix the builds.

**Comment by otto-md** [comment]
I just squashed the commits now. Thanks for the help and review.

File: langserver/format.go [hunk 6]
```diff
@@ -13,6 +13,7 @@ import (
 
 	"github.com/pmezard/go-difflib/difflib"
 	"golang.org/x/tools/go/buildutil"
+	"golang.org/x/tools/imports"
```
Comment (keegancsmith) [inline langserver/format.go]: i don't think this is in our vendor tree. Can you run `dep ensure`?

Comment (otto-md) [reply langserver/format.go]: It is: https://github.com/sourcegraph/go-langserver/tree/master/vendor/golang.org/x/tools/imports

File: README.md [hunk 7]
```diff
@@ -42,6 +42,20 @@ interface GoInitializationOptions {
    */
   gocodeCompletionEnabled?: boolean;
 
+  /**
+   * goimportsEnabled formats documents using goimports instead of normal go fmt.
+   *
+   * Defaults to true if not specified.
+   */
+  goimportsEnabled?: boolean;
```
[the remark this answers is not in the document]
Comment (otto-md) [reply README.md]: Sure: `formatTool?: string;` ?

that defaults to `"goimports"` but can also be `"gofmt"`?

And something like this in config.go?
```go
type FormatTool string

const (
	FormatToolGoimports FormatTool = "goimports"
	FormatToolGofmt     FormatTool = "gofmt"
)
```

[the remark this answers is not in the document]
Comment (keegancsmith) [reply README.md]: Yup that looks good to me. Thanks!

File: README.md [hunk 8]
```diff
@@ -43,11 +43,11 @@ interface GoInitializationOptions {
...
 
   /**
-   * goimportsEnabled formats documents using goimports instead of normal go fmt.
+   * formatTool decides which tool is used to format documents. Supported: goimports and gofmt.
    *
-   * Defaults to true if not specified.
+   * Defaults to goimports if not specified.
    */
-  goimportsEnabled?: boolean;
+  formatTool?: string;
```
Comment (keegancsmith) [inline README.md]: given this is typescript, you could write this using a union of string literal types. `formatTool?: "goimports" | "gofmt";`. I'll update once merged.

**Comment by keegancsmith** [comment]
Thanks so much!

**Comment by maximbaz** [comment]
Thank you both for getting this in `master`!

**Pull request merged**